src/selector-native.js

```diff
@@ -62,18 +74,59 @@ jQuery.extend( {
 				}
 			}
 		} else {
-			jQuery.merge( results, context.querySelectorAll( selector ) );
+
+			newSelector = selector;
+			newContext = context;
+
+			// qSA considers elements outside a scoping root when evaluating child or
+			// descendant combinators, which is not what we want.
+			// In such cases, we work around the behavior by prefixing every selector in the
+			// list with an ID selector referencing the scope context.
+			// The technique has to be used as well when a leading combinator is used
+			// as such selectors are not recognized by querySelectorAll.
+			// Thanks to Andrew Dupont for this technique.
+			if ( nodeType === 1 &&
+				( rdescend.test( selector ) || rcombinators.test( selector ) ) ) {
+
+				// Expand context for sibling selectors
+				newContext = rsibling.test( selector ) &&
+					testContext( context.parentNode ) ||
+					context;
+
+				// Outside of IE, if we're not changing the context we can
+				// use :scope instead of an ID.
+				if ( newContext !== context || isIE ) {
+
+					// Capture the context ID, setting it first if necessary
+					if ( ( nid = context.getAttribute( "id" ) ) ) {
+						nid = jQuery.escapeSelector( nid );
+					} else {
+						context.setAttribute( "id", ( nid = jQuery.expando ) );
```

Comment:
Where is this id attribute removed?

Reply:
Nice catch. I removed the `try-catch` since we don't do custom traversal here but I should have instead changed it into `try-finally` to keep this `finally` block:
https://github.com/jquery/jquery/blob/b02a257f98688aa890e06a85672cd1a54c3ffa3a/src/selector.js#L288-L290

Reply:
@gibson042 The `finally` block added.